Sources/ReadingListKit/ReadingListModel.swift

```diff
@@ -0,0 +1,168 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import Combine
+import MoSoAnalytics
+
+protocol ReadingListModelDelegate: AnyObject {
+    func didFetchReadingListItems(urlStrings: [String], totalItemCount: Int)
+    func removeItemFromList(item: String)
+    func operationFailed(with error: PALError)
+}
+
+enum ReadingListDisplayMode {
+    case normal
+    case empty
+    case error
+    case loggedOut
+}
+
+public class ReadingListModel: ReadingListModelDelegate, ObservableObject {
+    private let analytics: ReadingListTracker
+
+    var pocketAccessLayer: PocketAccessLayer
+    var totalNumberOfItemsInReadingList: Int?
+    @Published private(set) var displayMode: ReadingListDisplayMode = .normal
+    @Published private(set) var readingListItems: [ReadingListCellViewModel] = []
+
+    public init(sessionProvider: @escaping ReadingListSessionProvider, consumerKey: String, analyticsTracker: ReadingListTracker) {
+        self.analytics = analyticsTracker
+        pocketAccessLayer = PocketAccessLayerImplementation(sessionProvider, consumerKey)
+        pocketAccessLayer.delegate = self
+        pocketAccessLayer.initApolloClient()
+
+        fetchMoreReadingList()
+    }
+
+    // MARK: - Fetch Reading List Items
+
+    func didDisplay(item: ReadingListCellViewModel) {
+        if allItemsAreDownloaded() {
+            return
+        }
+
+        let loadMoreThreshold = readingListItems.count - 3
+        if readingListItems.firstIndex(of: item) == loadMoreThreshold {
+            fetchMoreReadingList()
+        }
+    }
+
+    func fetchMoreReadingList() {
+        pocketAccessLayer.fetchSaves()
+    }
+
+    func allItemsAreDownloaded() -> Bool {
+        totalNumberOfItemsInReadingList == readingListItems.count
+    }
+
+    // MARK: - Delegate Methods
+
+    func didFetchReadingListItems(urlStrings: [String], totalItemCount: Int) {
+        totalNumberOfItemsInReadingList = totalItemCount
+
+        displayMode = .normal
+
+        if totalItemCount == 0 {
+            displayMode = .empty
+        }
+
+        urlStrings.forEach { urlString in
+            Task {
+                guard let item = try? await pocketAccessLayer.getItemForURL(urlString) else { return }
+                let readingListItem = readingListItem(from: item)
+                await MainActor.run {
+                    readingListItems.append(readingListItem)
+                }
+            }
+        }
```

Comment:
Does the backend not give us the items in the correct order?

Reply:
I'm not positive… but it's best to verify what you're seeing on web or iOS vs what you're seeing in your code, just to double check. I remember sorting on iOS, but that might just be because we're fetching from Core Data and need to add a sort descriptor.

Reply:
I checked and it seems like the pages come down in the expected order, but sometimes the items in the page are reordered (In one instance I think a page was reversed?).

Good enough for now?